### DISCUSSION TITLE
How to Start Edge Creation by Clicking on a Port?

### DISCUSSION THREAD

**Discussion by fengwang913** [eclipse-glsp/glsp#1444] [Q&A]
Hi,
I'm currently working on a graph editing tool and I'm facing a challenge with the edge creation process. I'm using the TriggerEdgeCreationAction to initiate the edge creation, but it seems to be in a state where the source is not preselected. This doesn't fully meet my requirements.

My Goal:
I want to achieve a functionality where clicking on a port automatically selects the source, and then the user can simply wait to connect to a target. This would streamline the user experience significantly.

Current Implementation:

1. I have set up event listeners on the ports to trigger the edge creation process.
2. The TriggerEdgeCreationAction is called, but it doesn't preselect the source port as I'd like.

Question:
Does anyone have experience with preselecting the source when initiating edge creation through a port click? Are there any specific techniques or workarounds that I could use to achieve this?

Any insights or suggestions would be greatly appreciated. Thank you in advance for your help!

**Comment by martin-fleck-at** [comment]
Hi @fengwang913,

Is there a particular reason you need to set up your own listener on the click of the port? I believe that if you are using `GPort` which already is a connectable element, the [EdgeCreationTool](https://github.com/eclipse-glsp/glsp-client/blob/8de5d139f94af875d6c073bf8d89107020f7a64d/packages/client/src/features/tools/edge-creation/edge-creation-tool.ts#L209) (when enabled) would already take care of that for you but maybe your use case is a little bit different.

You are correct in that the current `TriggerEdgeCreationAction` only triggers the`EdgeCreationTool` to be activated and the tool then allows you to select the source and target object by clicking on two connectable elements. The good news is that you can customize this behavior. In your case you'd probably need to do something like this:
- Add the pre-selected source element id to the `args` of the `TriggerEdgeCreationAction` you are sending
- Overwrite the default `EdgeCreationTool` (and most notably the `EdgeCreationToolMouseListener`) with your own implementation so that you already set the [source property](https://github.com/eclipse-glsp/glsp-client/blob/8de5d139f94af875d6c073bf8d89107020f7a64d/packages/client/src/features/tools/edge-creation/edge-creation-tool.ts#L90) on creation based on the trigger action that is given in the constructor.

I hope that helps you get started in the right direction!

**Reply by fengwang913** [reply]
Hi @martin-fleck-at 
Thank you for your response and suggestions! I really appreciate your input.

I’ve successfully implemented the solution by customizing the EdgeCreationTool and EdgeCreationToolMouseListener to set the source element ID directly through the TriggerEdgeCreationAction. By passing the pre-selected source ID into the action arguments, I was able to streamline the edge creation process so that clicking on a port automatically selects it as the source. This made the user experience much smoother!

Here's a summary of what I did:

1. I modified the EdgeCreationToolMouseListener to accept the source as part of the TriggerEdgeCreationAction args.
2. I used the args.id to pre-select the source and trigger the feedback for the edge creation accordingly.
3. I ensured that the feedback was updated dynamically, allowing users to connect to a target element.

Thanks again for pointing me in the right direction! It worked out well for my use case.
